### PR TITLE
COL-496 - Implement md5 checking

### PR DESCRIPTION
**Pull request by ayang99**
- preserve part checksums during upload, and retrieve/compare during downloads

COL-495 - remove double-counting of bytes in upload progress

### CODE REVIEW

**Comment by ayang99** [comment]
retest

File: dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/download/DownloadService.java [hunk 1]
```diff
@@ -119,18 +119,20 @@ public String getUrlAsString(@NonNull String objectId, long offset, long length)
...
-  public void download(File outputDirectory, String objectId, long offset, long length, boolean redo)
+  public void download(File outputDirectory, String objectId, long offset, long length, boolean force)
       throws IOException {
     log.debug("Beginning download of {} to {} {} - {}", objectId, outputDirectory.toString(), offset, length);
     int retry = 0;
     for (; retry < retryNumber; retry++) {
       try {
-        if (redo) {
+        if (force) {
+          // create local file handle for output
```
Comment (btiernay) [inline dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/download/DownloadService.java]: Please [capitalize all comments](http://docs.icgc.org/software/development/#comments).

File: dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/download/DownloadStateStore.java [hunk 2]
```diff
@@ -107,11 +108,12 @@ public boolean hasStarted(File stateDir, String objectId) {
   public List<Part> getProgress(File stateDir, String objectId) throws IOException {
     log.debug("Loading local progress for {} from {}", objectId, stateDir.toString());
     ObjectSpecification spec = loadSpecification(stateDir, objectId);
-    log.debug("Completed loading local object specification");
+    log.debug("Completed loading local object specification (meta file)");
     for (Part part : spec.getParts()) {
       log.debug("Checking md5 for part {}", part.getPartNumber());
       if (isCompleted(stateDir, objectId, part)) {
         Part completedPart = loadPart(stateDir, objectId, getPartName(part));
+        // copy download md5 into ObjectSpecification
```
Comment (btiernay) [inline dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/download/DownloadStateStore.java]: Ditto.

File: dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/download/DownloadStateStore.java [hunk 3]
```diff
@@ -107,11 +108,12 @@ public boolean hasStarted(File stateDir, String objectId) {
   public List<Part> getProgress(File stateDir, String objectId) throws IOException {
     log.debug("Loading local progress for {} from {}", objectId, stateDir.toString());
     ObjectSpecification spec = loadSpecification(stateDir, objectId);
-    log.debug("Completed loading local object specification");
+    log.debug("Completed loading local object specification (meta file)");
```
Comment (btiernay) [inline dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/download/DownloadStateStore.java]: Might as well throw in `File stateDir, String objectId` to the message for good measure.

File: dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/transport/StorageService.java [hunk 4]
```diff
@@ -124,41 +124,43 @@ public void downloadPart(DataChannel channel, Part part, String objectId, File o
       public Void doWithRetry(RetryContext ctx) throws IOException {
         log.debug("Download Part URL: {}", part.getUrl());
 
-        final RequestCallback callback = new RequestCallback() {
+        final RequestCallback requestCallback = new RequestCallback() {
```
Comment (btiernay) [inline dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/transport/StorageService.java]: You might consider using Java 8 lambdas here and elsewhere to cut down on a lot of noise:

```java
          HttpHeaders headers =
              dataTemplate.execute(new URI(part.getUrl()), GET,

                  request -> request.getHeaders().set(RANGE, Parts.getHttpRangeValue(part)),

                  response -> {
                    try (HashingInputStream his = new HashingInputStream(Hashing.md5(), response.getBody())) {
                      channel.readFrom(his);
                      response.getHeaders().set(ETAG, his.hash().toString());
                      return response.getHeaders();
                    }
                  });
;
```

File: dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/transport/StorageService.java [hunk 5]
```diff
@@ -124,41 +124,43 @@ public void downloadPart(DataChannel channel, Part part, String objectId, File o
...
           // the actual GET operation
           log.debug("performing GET {}", part.getUrl());
-          HttpHeaders headers =
-              dataTemplate.execute(new URI(part.getUrl()), HttpMethod.GET, callback, headersExtractor);
-          part.setMd5(cleanUpETag(headers.getETag()));
+          val md5 = dataTemplate.execute(new URI(part.getUrl()), HttpMethod.GET, requestCallback, headersExtractor);
+          part.setMd5(md5);
+          if (part.hasFailedChecksum()) {
+            val msg = String.format("Checksum failed for Part# %d: %s", part.getPartNumber(), part.getMd5());
+            throw new Exception(msg);
```
Comment (btiernay) [inline dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/transport/StorageService.java]: use `checkState` instead. Not only does it throw amore informative exception but it reads better and removes a branch:

```java
checkState(!part.hasFailedChecksum(), "Checksum failed for Part# %d: %s", part.getPartNumber(), part.getMd5()); 
```

File: dcc-storage-core/src/main/java/org/icgc/dcc/storage/core/model/Part.java [hunk 6]
```diff
@@ -48,4 +49,8 @@ public int compareTo(Part otherPart) {
     return this.partNumber - otherPart.partNumber;
   }
 
+  @JsonIgnore
+  public boolean hasFailedChecksum() {
+    return !sourceMd5.equals(md5);
+  }
```
Comment (btiernay) [inline dcc-storage-core/src/main/java/org/icgc/dcc/storage/core/model/Part.java]: Add new line

File: dcc-storage-server/src/main/java/org/icgc/dcc/storage/server/controller/ObjectUploadController.java [hunk 7]
```diff
@@ -114,7 +115,10 @@ public void finalizeUpload(
       @PathVariable(value = "object-id") String objectId,
       @RequestParam(value = "uploadId", required = true) String uploadId
       ) {
+    val startMs = System.currentTimeMillis();
```
Comment (btiernay) [inline dcc-storage-server/src/main/java/org/icgc/dcc/storage/server/controller/ObjectUploadController.java]: Please use `Stopwatch` for this as it is cleaner and the output is much more readable for humans:

```java
val watch = Stopwatch.createStarted();
//...
log.info("Finished ... in {}", watch);
```

File: dcc-storage-server/src/main/java/org/icgc/dcc/storage/server/service/upload/ObjectUploadService.java [hunk 8]
```diff
@@ -254,20 +257,29 @@ public void finalizeUploadPart(String objectId, String uploadId, int partNumber,
...
 
     val actualBucketName = bucketNamingService.getObjectBucketName(objectId);
     val actualStateBucketName = bucketNamingService.getStateBucketName(objectId);
 
     if (stateStore.isCompleted(objectId, uploadId)) {
       try {
-        val etags = stateStore.getUploadStatePartEtags(objectId, uploadId);
+        val details = stateStore.getUploadStatePartDetails(objectId, uploadId);
+        val etags = Lists.<PartETag> newArrayList();
+        details.values().forEach(detail -> etags.add(detail.getEtag()));
```
Comment (btiernay) [inline dcc-storage-server/src/main/java/org/icgc/dcc/storage/server/service/upload/ObjectUploadService.java]:
```java
import static org.icgc.dcc.common.core.util.stream.Streams.toImmutableList
...
val etags = details.values().stream().map(detail -> detail.getETag()).collect(toImmutableList())
```

File: dcc-storage-server/src/main/java/org/icgc/dcc/storage/server/service/upload/UploadStateStore.java [hunk 9]
```diff
@@ -93,6 +96,15 @@
   @Autowired
   private BucketNamingService bucketNamingService;
 
+  @Data
+  @NoArgsConstructor
+  class UploadPartDetail {
+
+    private PartETag etag;
+    private int partNumber;
+    private String md5;
```
Comment (btiernay) [inline dcc-storage-server/src/main/java/org/icgc/dcc/storage/server/service/upload/UploadStateStore.java]: New line

File: dcc-storage-server/src/main/java/org/icgc/dcc/storage/server/service/upload/UploadStateStore.java [hunk 10]
```diff
@@ -93,6 +96,15 @@
   @Autowired
   private BucketNamingService bucketNamingService;
 
+  @Data
+  @NoArgsConstructor
```
Comment (btiernay) [inline dcc-storage-server/src/main/java/org/icgc/dcc/storage/server/service/upload/UploadStateStore.java]: I think `@NoArgsConstructor` is implied by `@Data`. In general, please prefer `@Value`. If you need to mutate, consider `@Wither`

File: dcc-storage-server/src/main/java/org/icgc/dcc/storage/server/service/upload/UploadStateStore.java [hunk 11]
```diff
@@ -278,18 +290,21 @@ public void finalizeUploadPart(String objectId, String uploadId, int partNumber,
...
-    val etags = Lists.<PartETag> newArrayList();
+    val details = Maps.<Integer, UploadPartDetail> newHashMap();
 
     eachObjectSummary(objectId, uploadStateKey, (objectSummary) -> {
       CompletedPart part = readCompletedPart(objectId, uploadId, objectSummary);
+      UploadPartDetail detail = new UploadPartDetail();
       PartETag etag = new PartETag(part.getPartNumber(), part.getEtag());
-
-      etags.add(etag);
+      detail.setEtag(etag);
```
Comment (btiernay) [inline dcc-storage-server/src/main/java/org/icgc/dcc/storage/server/service/upload/UploadStateStore.java]: Ya `@Value` with `@Builder` seems more appropriate. Better thread safety and invariant guarantees.

File: dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/transport/MemoryMappedParallelPartObjectTransport.java [hunk 12]
```diff
@@ -115,13 +115,11 @@ public Part call() throws Exception {
                   log.debug("Fail checksum. Reupload part: {}", part);
                   progress.startTransfer();
                   proxy.uploadPart(channel, part, objectId, uploadId);
-                  progress.incrementBytesWritten(part.getPartSize());
```
Comment (btiernay) [inline dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/transport/MemoryMappedParallelPartObjectTransport.java]: Where does this happen now?

File: dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/transport/MemoryMappedParallelPartObjectTransport.java [hunk 13]
```diff
@@ -115,13 +115,11 @@ public Part call() throws Exception {
...
                 }
                 progress.incrementChecksumParts();
               } else {
                 progress.startTransfer();
                 proxy.uploadPart(channel, part, objectId, uploadId);
-                progress.incrementBytesWritten(part.getPartSize());
```
Comment (btiernay) [inline dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/transport/MemoryMappedParallelPartObjectTransport.java]: Where does this happen now?

File: dcc-storage-server/src/main/java/org/icgc/dcc/storage/server/service/upload/UploadPartDetail.java [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+
+import com.amazonaws.services.s3.model.PartETag;
+
+@Builder
+@Value
+public class UploadPartDetail {
+
+  private PartETag etag;
+  private int partNumber;
+  private String md5;
```
Comment (btiernay) [inline dcc-storage-server/src/main/java/org/icgc/dcc/storage/server/service/upload/UploadPartDetail.java]: Newline

File: dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/download/DownloadStateStore.java (hunk 3 above)
Comment (btiernay) [inline dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/download/DownloadStateStore.java]: I still think this is a good idea.

File: dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/transport/MemoryMappedParallelPartObjectTransport.java (hunk 13 above)
Comment (btiernay) [inline dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/transport/MemoryMappedParallelPartObjectTransport.java]: @ayang99  Are you sure about this change?

Comment (ayang99) [inline dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/transport/MemoryMappedParallelPartObjectTransport.java]: I've been regression-testing this part this past little while. It was calling incrementBytesWritten() here as well as in ProgressInputStream/ProgressOutputStream on every read/write call. This seemed like double-counting to me.

File: dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/download/DownloadStateStore.java (hunk 3 above)
Comment (ayang99) [inline dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/download/DownloadStateStore.java]: it's already logged two lines earlier?

Comment (btiernay) [inline dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/download/DownloadStateStore.java]: Touche!

File: dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/transport/MemoryMappedParallelPartObjectTransport.java (hunk 13 above)
Comment (btiernay) [inline dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/transport/MemoryMappedParallelPartObjectTransport.java]: Ok, keep in mind that the ProgressInputStream is shared between implementations of transport. So you have to ensure that the other implementations continue to work correctly.

**Comment by btiernay** [comment]
Please address merge conflicts

**Comment by btiernay** [comment]
@ayang99 Please address merge conflicts

**Pull request merged**